Skip adding alias targets when they don't exist

porglezomp opened the pull request:
The CMake uses the concept of a "primary variant" which isn't necessarily aligned with either the host or target. In some cases, like cross-compiling an iOS compiler toolchain for `macosx-arm64`, the expected "primary variant" target will be missing and so CMake will fail. We can skip adding the alias since `build-script` will call the more specific target anyway.

porglezomp committed:
Skip adding alias targets when they don't exist

The CMake uses the concept of a "primary variant" which isn't
necessarily aligned with either the host or target. In some cases, like
cross-compiling an iOS compiler toolchain for macosx-arm64, the expected
"primary variant" target will be missing and so CMake will fail. We can
skip adding the alias since build-script will call the more specific
target anyway.

porglezomp commented:
@swift-ci please smoke test

edymtt commented on stdlib/CMakeLists.txt:
```diff
@@ -196,13 +196,23 @@ function(swift_create_stdlib_targets name variant define_all_alias)
+  # so it can't be added as a target here. build-script will invoke the
+  # more-specific target, so just skip creating this target and warn in case
+  # someone is using the CMake more directly.
+  if(TARGET ${name}${SWIFT_PRIMARY_VARIANT_SUFFIX}${variant})
```
I believe a more readable check would be

```suggestion
  if(SWIFT_PRIMARY_VARIANT IN_LIST SWIFT_SDKS)
```

edymtt commented on stdlib/CMakeLists.txt:
```diff
@@ -196,13 +196,23 @@ function(swift_create_stdlib_targets name variant define_all_alias)
+      PROPERTIES
+      FOLDER "Swift libraries/Aggregate")
+  else()
+    message(WARNING "Primary variant is not being built, target ${name}${SWIFT_PRIMARY_VARIANT_SUFFIX}${variant} does not exist, not creating ${name}${variant} alias target.")
```
If we decide to change the check, we could improve on this message as well, e.g.

```suggestion
    message(WARNING "Primary variant ${SWIFT_PRIMARY_VARIANT} is not being built, not creating ${name}${variant} alias target for it.")
```

edymtt replied:
(Optional) would it be possible to add some tests for this scenario?

porglezomp committed:
[cmake] Skip primary variant alias in a cleaner way

Adjust the condition and message to skip the primary variant alias in a
way that's more clear and more directly matches the explanation in the
comments.

Co-authored-by: Eric Miotto <emiotto@apple.com>

porglezomp commented:
@swift-ci please smoke test

porglezomp commented:
I'm also rebuilding locally against a preset that I'm sure exercises cross-compilation

porglezomp commented:
@swift-ci please smoke test macOS Platform

porglezomp commented:
@swift-ci please smoke test macOS Platform

edymtt commented on stdlib/CMakeLists.txt:
```diff
@@ -196,13 +196,23 @@ function(swift_create_stdlib_targets name variant define_all_alias)
+  # so it can't be added as a target here. build-script will invoke the
+  # more-specific target, so just skip creating this target and warn in case
+  # someone is using the CMake more directly.
+  if(SWIFT_PRIMARY_VARIANT IN_LIST SWIFT_SDKS)
```
I fat fingered a variable name in my previous suggestion, this should be the correct one

```suggestion
  if(SWIFT_PRIMARY_VARIANT_SDK IN_LIST SWIFT_SDKS)
```

edymtt commented on stdlib/CMakeLists.txt:
```diff
@@ -196,13 +196,23 @@ function(swift_create_stdlib_targets name variant define_all_alias)
+      PROPERTIES
+      FOLDER "Swift libraries/Aggregate")
+  else()
+    message(WARNING "Primary variant ${SWIFT_PRIMARY_VARIANT} is not being built, not creating ${name}${variant} alias target for it.")
```
```suggestion
    message(WARNING "Primary variant ${SWIFT_PRIMARY_VARIANT_SDK} is not being built, not creating ${name}${variant} alias target for it.")
```

compnerd reviewed (approved):
I believe that this should be safe.

porglezomp commented:
Just squashing the two review fixup commits and improving their commit message.

porglezomp commented:
@swift-ci please smoke test and merge

porglezomp commented:
@swift-ci please smoke test and merge

porglezomp commented:
@swift-ci please smoke test macOS Platform and merge

porglezomp commented:
@swift-ci please smoke test macOS Platform

porglezomp commented:
@swift-ci please smoke test macOS Platform
